modules/platforms/cpp/binary/src/impl/binary/binary_reader_impl.cpp

```diff
@@ -795,6 +823,166 @@ namespace ignite
+                {
+                    int32_t realLen = stream->ReadInt32();
+                    if (realLen > 0)
+                        stream->Shift(realLen * 8);
+
+                    return;
+                }
+
+                case IGNITE_TYPE_ARRAY_STRING:
+                case IGNITE_TYPE_ARRAY_UUID:
+                case IGNITE_TYPE_ARRAY_DATE:
+                case IGNITE_TYPE_ARRAY_TIMESTAMP:
+                case IGNITE_TYPE_ARRAY_TIME:
+                case IGNITE_TYPE_ARRAY:
+                {
+                    int32_t cnt = stream->ReadInt32();
+                    for (int32_t i = 0; i < cnt; i++)
+                        Skip();
+
+                    return;
+                }
+
+                case IGNITE_TYPE_COLLECTION:
+                {
+                    int32_t cnt = stream->ReadInt32();
+
+                    // Collection type ID.
+                    stream->Shift(1);
+
+                    for (int32_t i = 0; i < cnt; i++)
+                        Skip();
+
+                    return;
+                }
+
+                case IGNITE_TYPE_MAP:
+                {
+                    int32_t cnt = stream->ReadInt32();
+
+                    // Map type ID.
+                    stream->Shift(1);
+
+                    for (int32_t i = 0; i < cnt; i++)
+                    {
+                        Skip();
+                        Skip();
+                    }
+
+                    return;
+                }
+
+                case IGNITE_TYPE_TIMESTAMP:
+                    stream->Shift(12);
+                    return;
+
+                case IGNITE_HDR_FULL:
+                    // Get position back to IGNITE_HDR_FULL
+                    stream->Position(stream->Position() - 1);
+                    EmptyType val;
+                    ReadTopObject0<EmptyReader, EmptyType>(val);
+                    return;
+
+                case IGNITE_HDR_NULL:
+                    return;
+
+                default:
+                {
+                    int32_t pos = stream->Position() - 1;
+
+                    switch (typeId)
+                    {
+                    case IGNITE_TYPE_MAP_ENTRY:
+                    case IGNITE_TYPE_BINARY:
+                    case IGNITE_TYPE_ENUM:
+                    case IGNITE_TYPE_ARRAY_ENUM:
+                    case IGNITE_TYPE_DECIMAL:
+                    case IGNITE_TYPE_ARRAY_DECIMAL:
+                    case IGNITE_TYPE_CLASS:
+                    case IGNITE_TYPE_PROXY:
+                    case IGNITE_TYPE_BINARY_ENUM:
+                        IGNITE_ERROR_FORMATTED_2(IgniteError::IGNITE_ERR_BINARY, "Invalid header", "position", pos,
+                            "unsupported type", static_cast<int>(typeId))
+                    default:
+                        IGNITE_ERROR_FORMATTED_2(IgniteError::IGNITE_ERR_BINARY, "Invalid header", "position", pos,
+                            "invalid type", static_cast<int>(typeId))
```

Comment:
I believe, it should be _"unsupported type"_ for all cases, as we can introduce new types in future, which are not _invalid_.

Reply:
Fixed.